lib/classes/ConfigSchemaHandler.js

```diff
@@ -0,0 +1,76 @@
+'use strict';
+
+const Ajv = require('ajv');
+const schema = require('../configSchema');
+const ajvErrorMessageBuilder = require('../utils/ajvErrorMessageBuilder');
+
+const functionNamePattern = '^[a-zA-Z0-9-_]+$';
+
+class ConfigSchemaHandler {
+  constructor(serverless) {
+    this.serverless = serverless;
+    this.schema = schema;
+
+    deepFreeze(this.schema.properties.service);
+    deepFreeze(this.schema.properties.plugins);
+    deepFreeze(this.schema.properties.resources);
+    deepFreeze(this.schema.properties.package);
+    deepFreeze(this.schema.properties.layers);
+    deepFreeze(this.schema.properties.outputs);
```

Comment:
This I believe should not be part of schema at all.

And I guess list of top (global) properties should be froze (so no new properties in main context are allowed)

Reply:
I froze global properties at the end of `defineProvider` method because it did not let make necessary extensions